include/core/lf_token.h

```diff
@@ -0,0 +1,362 @@
+/**
+ * @file
+ * @author Edward A. Lee (eal@berkeley.edu)
+ *
+ * @section LICENSE
+ * Copyright (c) 2022, The University of California at Berkeley.
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @section DESCRIPTION
+ *
+ * This header file supports token objects, which are reference-counted wrappers
+ * around values that are carried by events scheduled on the event queue and held
+ * in ports and actions when the type is not a primitive type.
+ * 
+ * A token has type lf_token_t. It points to a value, a dynamically allocated
+ * chunk of memory on the heap. It has a length field, which enables its value
+ * to be interpreted as an array of the given length. It has a pointer to type
+ * (token_type_t), which has an element_size field specifying the size of each
+ * array element (or the size of the whole value if it is not an array and has
+ * length 1). The type also optionally has function pointers to a destructor
+ * and copy constructor. These must be specified if the payload (value) is a complex
+ * struct that cannot be freed by a simple call to free() or copied by a call
+ * to memcpy().
+ * 
+ * An instance of a port struct and trigger_t struct (an action or an input port)
+ * can be cast to token_template_t, which has a token_type_t field called type
+ * and a pointer to a token (which may be NULL).  The same instance can also be
+ * cast to token_type_t, which has an element_size field and (possibly) function
+ * pointers to a destructor and a copy constructor.
+ * 
+ * A "template token" is one pointed to by a token_template_t (an action or a port).
```

Comment:
It's still not really clear to me what a template token _is_ (or why we have it) after reading this text.

Reply:
It looks like it supports the persistence of port values until they are overwritten, as Edward has been advocating for (?)

Reply:
I have improved this documentation and pushed to the examples repository an [implementation of the ROSACE case study](https://github.com/lf-lang/examples-lingua-franca/blob/main/C/src/periodic/README.md) (an aircraft controller) that requires persistence of inputs. This example would be much more complicated, I think, in the Cpp target, which does not support persistent inputs.